NPE in EqualsAvoidNull check when accessing string from base class with 'this' qualifier

Checkstyle ends with a NPE when a derived class accesses an String variable from a base class when the variable is accessed via 'this' qualifier. The check runs fine when the 'this.' is omitted.

This error appeared first in the Checkstyle version 8.29 we currently use but can be reprodued with the latest 9.1 version. 
It appears on the checkstyle command line (as in the sample below), withing the checkstyle maven plugin and in the checkstyle Eclipse plugin.

**The below command sequence demonstrates the error:**
```
C:\Temp\test>type AbstractBaseClass.java 
package com.fsp.test;

public abstract class AbstractBaseClass
{
   protected String stringFromBaseClass = "ABC";
}

C:\Temp\test>type DerivedClass.java 
package com.fsp.test;

public class DerivedClass extends AbstractBaseClass
{
   public DerivedClass()
   {
      if ( this.stringFromBaseClass.equals("JKHKJ") ) System.out.println("Hello");
   }
}

C:\Temp\test>type checkstyle-config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="fileExtensions" value="java"/>
  <module name="TreeWalker">
    <module name="EqualsAvoidNull"/>
  </module>
</module>

C:\Temp\test>java -Duser.language=en -Duser.country=US -jar c:\temp\checkstyle-9.1-all.jar -c checkstyle-config.xml  AbstractBaseClass.java DerivedClass.java 
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing DerivedClass.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:302)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:409)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:332)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:191)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:126)
Caused by: java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.getFieldType(EqualsAvoidNullCheck.java:545)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.isStringFieldOrVariableFromThisInstance(EqualsAvoidNullCheck.java:481)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.isCalledOnStringFieldOrVariable(EqualsAvoidNullCheck.java:439)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.checkMethodCall(EqualsAvoidNullCheck.java:356)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.traverseFieldFrameTree(EqualsAvoidNullCheck.java:339)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.traverseFieldFrameTree(EqualsAvoidNullCheck.java:336)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.traverseFieldFrameTree(EqualsAvoidNullCheck.java:336)
	at com.puppycrawl.tools.checkstyle.checks.coding.EqualsAvoidNullCheck.finishTree(EqualsAvoidNullCheck.java:242)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyEnd(TreeWalker.java:319)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:274)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:154)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:328)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:289)
	... 5 more
Checkstyle ends with 1 errors.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheck.java
@@ -478,7 +478,7 @@ public class EqualsAvoidNullCheck extends AbstractCheck {
     private boolean isStringFieldOrVariableFromThisInstance(DetailAST objCalledOn) {
         final String name = objCalledOn.getText();
         final DetailAST field = getObjectFrame(currentFrame).findField(name);
-        return STRING.equals(getFieldType(field));
+        return field != null && STRING.equals(getFieldType(field));
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheckTest.java
@@ -179,6 +179,16 @@ public class EqualsAvoidNullCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
+    public void testEqualsSuperClass() throws Exception {
+
+        final String[] expected = {
+            "23:35: " + getCheckMessage(MSG_EQUALS_AVOID_NULL),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputEqualsAvoidNullSuperClass.java"), expected);
+    }
+
+    @Test
     public void testInputEqualsAvoidNullEnhancedInstanceof() throws Exception {
 
         final String[] expected = {

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/equalsavoidnull/InputEqualsAvoidNullSuperClass.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/equalsavoidnull/InputEqualsAvoidNullSuperClass.java
@@ -1,0 +1,26 @@
+/*
+EqualsAvoidNull
+ignoreEqualsIgnoreCase = (default)false
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.coding.equalsavoidnull;
+
+public abstract class InputEqualsAvoidNullSuperClass {
+   protected String stringFromBaseClass = "ABC";
+}
+
+class DerivedClass extends InputEqualsAvoidNullSuperClass {
+   protected String classField = "DEF";
+
+    void m1() {
+        if (this.stringFromBaseClass.equals("JKHKJ")) { // ok
+        }
+    }
+
+    void m2() {
+        if (this.classField.equals("JKHKJ")) { // violation
+        }
+    }
+}

